File: src/main/java/gay/skitbet/jackiro/Jackiro.java
package gay.skitbet.jackiro;

import gay.skitbet.jackiro.managers.CommandHandler;
import gay.skitbet.jackiro.managers.MongoManager;
import gay.skitbet.jackiro.database.ServerConfigRepository;
import gay.skitbet.jackiro.listener.MainListener;
import gay.skitbet.jackiro.music.JackiroMusicManager;
import gay.skitbet.jackiro.task.LoadGuildsTask;
import gay.skitbet.jackiro.task.UpdateStatusTask;
import gay.skitbet.jackiro.utils.JackiroConfig;
import lombok.Getter;
import net.dv8tion.jda.api.OnlineStatus;
import net.dv8tion.jda.api.entities.Activity;
import net.dv8tion.jda.api.requests.GatewayIntent;
import net.dv8tion.jda.api.sharding.DefaultShardManagerBuilder;
import net.dv8tion.jda.api.sharding.ShardManager;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.Random;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;

@Getter
public class Jackiro {

    private static Jackiro instance;

    // random utils
    public static final Random RANDOM = new Random();
    public static final Logger LOGGER = LoggerFactory.getLogger(Jackiro.class);
    public static JackiroConfig config;

    // important bot stuff
    private final ScheduledExecutorService executor;
    private final Instant startTime = Instant.now();
    private JackiroMusicManager jackiroMusicManager;
    private CommandHandler commandHandler;

    // db repositories
    private ServerConfigRepository serverConfigRepository;

    // jda shard manager
    public int readyShards;
    private ShardManager shardManager;


    public Jackiro(JackiroConfig config) {
        instance = this;
        Jackiro.config = config; // load config early
        this.executor = Executors.newScheduledThreadPool(8); // create async executor
    }

    /**
     * Gets the Jackiro instance.
     * @return Jackiro instance
     */
    public static Jackiro getInstance() {
        if (instance == null) {
            throw new IllegalStateException("Jackiro has not been initialized yet.");
        }
        return instance;
    }

    /**
     * Boots up the bot and starts everything.
     * @param config config to use
     */
    public static void initialize(JackiroConfig config) {
        if (instance != null) {
            throw new IllegalStateException("Jackiro has already been initialized.");
        }
        new Jackiro(config).start();
    }

    /**
     * Starts up the bot (db, shards, etc).
     */
    private void start() {
        try {
            initializeMongoDB(); // connect to mongo
            initializeShardManager(); // start shards
            this.commandHandler = new CommandHandler(shardManager);
            LOGGER.info("Jackiro bot has started successfully!");
        } catch (Exception e) {
            LOGGER.error("Failed to start Jackiro: ", e);
            shutdown();
        }
    }

    /**
     * Connects to MongoDB database.
     */
    private void initializeMongoDB() {
        LOGGER.info("Connecting to MongoDB...");
        MongoManager.connect("jackiro"); // connect to db
        this.serverConfigRepository = new ServerConfigRepository(); // setup repo
        LOGGER.info("MongoDB connected.");
    }

    /**
     * Initializes JDA's ShardManager and loads shards.
     */
    private void initializeShardManager() {
        LOGGER.info("Initializing Shard Manager...");
        DefaultShardManagerBuilder builder = DefaultShardManagerBuilder.createDefault(config.getClientToken())
                .setShardsTotal(config.getShardCount())
                .enableIntents(GatewayIntent.GUILD_VOICE_STATES, GatewayIntent.GUILD_MESSAGES, GatewayIntent.MESSAGE_CONTENT)
                .setActivity(Activity.listening("beep boop .. loading")) // funny starting status
                .setStatus(OnlineStatus.DO_NOT_DISTURB)
                .addEventListeners(new MainListener()); //  register main listener

        shardManager = builder.build();

    }

    /**
     * Called when shards are ready.
     * Starts background tasks like status updates and guild loading.
     */
    public void onReady() {
        new UpdateStatusTask().start(); // auto update status
        new LoadGuildsTask(this).start(); // cache guild settings
        this.commandHandler.registerCommands();
        this.jackiroMusicManager = new JackiroMusicManager();
    }

    /**
     * Shuts down the bot and kills everything safely.
     */
    public void shutdown() {
        LOGGER.warn("Shutting down Jackiro...");
        if (shardManager != null) {
            shardManager.shutdown(); // shutdown shards
        }
        executor.shutdownNow(); // stop async tasks
        System.exit(1); // peace out
    }
}

File: src/main/kotlin/gay/skitbet/jackiro/Jackiro.kt
package gay.skitbet.jackiro

import gay.skitbet.jackiro.database.ServerConfigRepository
import gay.skitbet.jackiro.listener.MainListener
import gay.skitbet.jackiro.managers.CommandHandler
import gay.skitbet.jackiro.managers.MongoManager
import gay.skitbet.jackiro.music.JackiroMusicManager
import gay.skitbet.jackiro.task.LoadGuildsTask
import gay.skitbet.jackiro.task.UpdateStatusTask
import gay.skitbet.jackiro.utils.JackiroConfig
import net.dv8tion.jda.api.OnlineStatus
import net.dv8tion.jda.api.entities.Activity
import net.dv8tion.jda.api.requests.GatewayIntent
import net.dv8tion.jda.api.sharding.DefaultShardManagerBuilder
import net.dv8tion.jda.api.sharding.ShardManager
import org.slf4j.LoggerFactory
import java.time.Instant
import java.util.*
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService

class Jackiro private constructor(config: JackiroConfig) {

    companion object {
        val random: Random = Random()
        val logger = LoggerFactory.getLogger(Jackiro::class.java)

        lateinit var instance: Jackiro
        lateinit var config: JackiroConfig
            private set

        fun getInstance(): Jackiro =
            instance ?: throw IllegalStateException("Jackiro has not been initialized yet.")

        fun initialize(config: JackiroConfig) {
            if (instance != null) {
                throw IllegalStateException("Jackiro has already been initialized.")
            }
            Jackiro(config).start()
        }
    }


    val executor: ScheduledExecutorService = Executors.newScheduledThreadPool(8)
    val startTime: Instant = Instant.now()
    var readyShards: Int = 0

    lateinit var jackiroMusicManager: JackiroMusicManager
        private set
    lateinit var commandHandler: CommandHandler
        private set
    lateinit var shardManager: ShardManager
        private set

    init {
        Companion.config = config
        instance = this
    }

    private fun start() {
        try {
            initializeMongoDB()
            initializeShardManager()
            commandHandler = CommandHandler(shardManager!!)
            logger.info("Jackiro bot has started successfully!")
        } catch (e: Exception) {
            logger.error("Failed to start Jackiro: ", e)
            shutdown()
        }
    }

    private fun initializeMongoDB() {
        logger.info("Connecting to MongoDB...")
        MongoManager.connect("jackiro")
        logger.info("MongoDB connected.")
    }

    private fun initializeShardManager() {
        logger.info("Initializing Shard Manager...")
        val builder = DefaultShardManagerBuilder.createDefault(config!!.clientToken)
            .setShardsTotal(config!!.shardCount)
            .enableIntents(
                GatewayIntent.GUILD_VOICE_STATES,
                GatewayIntent.GUILD_MESSAGES,
                GatewayIntent.MESSAGE_CONTENT
            )
            .setActivity(Activity.listening("beep boop .. loading"))
            .setStatus(OnlineStatus.DO_NOT_DISTURB)
            .addEventListeners(MainListener())

        shardManager = builder.build()
    }

    fun onReady() {
        UpdateStatusTask().start(executor)
        LoadGuildsTask(this).start(executor)
        commandHandler?.registerCommands()
        jackiroMusicManager = JackiroMusicManager()
    }

    fun shutdown() {
        logger.warn("Shutting down Jackiro...")
        shardManager?.shutdown()
        executor.shutdownNow()
        System.exit(1)
    }

}
